### DISCUSSION TITLE
Spectator showing one player running slow

### DISCUSSION THREAD

**Discussion by BroTon717** [ppy/osu#19429] [Q&A] [answered]
### Type

Game behaviour

### Bug description

Some players, myself included, have been experiencing weird score delays. The Live Play also have this problem with the addition of being very laggy. The delay became so bad that previous map scores are still being processed well into a new map.
(The log is from the player that experience this glitch.)

### Screenshots or videos

https://user-images.githubusercontent.com/110118379/181299321-1e62bdd2-06ec-468d-9fe2-edbc1d416d57.mp4

### Version

2022.723.0

### Logs

[network.log](https://github.com/ppy/osu/files/9201222/network.log)

**Comment by peppy** [comment, marked as answer]
We'd probably need a capture / full logs from the user experiencing this. Sounds like their system was lagging anyway? Which would make sense?

**Reply by Nepp0** [reply]
my system performance is fine like always, this "delay" issue has started from around a week ago, shortly after the latest update, i'm not sure if the infrastructure stuff is causing it to happen but ive noticed that the intensity of the delay varies from players, and doesn't get fixed after restarting the game, pc or router.
capture: https://youtu.be/4dA0ck-Yuw8
logs: 
[database.log](https://github.com/ppy/osu/files/9214795/database.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/9214796/legacy-ipc.log)
[network.log](https://github.com/ppy/osu/files/9214797/network.log)
[performance.log](https://github.com/ppy/osu/files/9214799/performance.log)
[performance-audio.log](https://github.com/ppy/osu/files/9214800/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/9214801/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/9214802/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/9214803/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/9214804/runtime.log)
[updater.log](https://github.com/ppy/osu/files/9214805/updater.log)

**Reply by peppy** [reply]
What are you showing in the capture by pointing at the leaderboard?

Also can you reproduce this? The spectator display looks to have given up keeping players in sync because one player was running behind, which is normal behaviour.

I can't reproduce this in multiple spectator sessions, so I'd wager on it being something local to your connection/routing/game. Hard to really make any further calls.

**Reply by peppy** [reply]
I was able to spectate a game, and observe this happen. @Nepp0's client is falling behind in sending data as his internet connection is unable to send the data in realtime. Probably due to using a VPN which is doing weird stuff.

We probably need better messaging on the client side to let the user know when their internet connection is broken, and we need better invalidation server-side to end a player's session if they fall too far behind like this.

**Reply by peppy** [reply]
In further discussion with @smoogipoo, I remembered that we recently made a change to sending of signalr packets which may have made this worse for cases of already high network latency (see https://github.com/ppy/osu/pull/19187/commits/0533249d11ac220e223369f774fafcabc3f30c51). I believe the change causes any potential batching/pipelining optimisations at signalr / websocket implementations to be lost, as we are waiting on a response for each frame bundle sent.

We'll revert this for the next release for the sending of spectator frames and see if it helps.

**Reply by peppy** [reply]
#19451
